readme/memcache.md

```diff
@@ -2,76 +2,198 @@
+ */
+
 if ($is_ah_env) {
   switch ($ah_env) {
     case 'test':
     case 'prod':
-      if ($modules && isset($modules['module']['memcache'])) {
-        // Use Memcached extension.
-        $memcached_exists = class_exists('Memcached', FALSE);
-        if ($memcached_exists) {
-          $settings['memcache']['extension'] = 'Memcached';
-        }
-
-        // Use memcache as the default bin.
-        $settings['cache']['default'] = 'cache.backend.memcache';
-
-        // Enable stampede protection.
-        $settings['memcache']['stampede_protection'] = TRUE;
-        // Move locks to memcache.
-        $settings['container_yamls'][] = DRUPAL_ROOT . '/../vendor/acquia/blt/settings/memcache.yml';
-      }
+
+  if (class_exists(\Composer\Autoload\ClassLoader::class)) {
+
+  $class_loader = new \Composer\Autoload\ClassLoader();
+  $class_loader->addPsr4('Drupal\\memcache\\', 'modules/contrib/memcache/src');
+
+  $class_loader->register();
+
+  $settings['container_yamls'][] = DRUPAL_ROOT . '/modules/contrib/memcache/memcache.services.yml';
+  
+   // Bootstrap cache.container with memcache rather than database.
+
+  $settings['bootstrap_container_definition'] = [
+    'parameters' => [],
+    'services' => [
+      'database' => [
+        'class' => 'Drupal\Core\Database\Connection',
+        'factory' => 'Drupal\Core\Database\Database::getConnection',
+        'arguments' => ['default'],
+      ],
+      'settings' => [
+        'class' => 'Drupal\Core\Site\Settings',
+        'factory' => 'Drupal\Core\Site\Settings::getInstance',
+      ],
+      'memcache.config' => [
+        'class' => 'Drupal\memcache\DrupalMemcacheConfig',
+        'arguments' => ['@settings'],
+      ],
+      'memcache.backend.cache.factory' => [
+        'class' => 'Drupal\memcache\DrupalMemcacheFactory',
+        'arguments' => ['@memcache.config']
+      ],
+      'memcache.backend.cache.container' => [
+        'class' => 'Drupal\memcache\DrupalMemcacheFactory',
+        'factory' => ['@memcache.backend.cache.factory', 'get'],
+        'arguments' => ['container'],
+      ],
+      'lock.container' => [
+        'class' => 'Drupal\memcache\Lock\MemcacheLockBackend',
+        'arguments' => ['container', '@memcache.backend.cache.container'],
+      ],
+      'cache_tags_provider.container' => [
+        'class' => 'Drupal\Core\Cache\DatabaseCacheTagsChecksum',
+        'arguments' => ['@database'],
+      ],
+      'cache.container' => [
+        'class' => 'Drupal\memcache\MemcacheBackend',
+        'arguments' => ['container', '@memcache.backend.cache.container', '@lock.container', '@memcache.config', '@cache_tags_provider.container'],
+      ],
+    ],
+  ];
+
+      // Override default fastchained backend for static bins. 
+      // See https://www.drupal.org/node/2754947
+
+      $settings['cache']['bins']['bootstrap'] = 'cache.backend.memcache';
+      $settings['cache']['bins']['discovery'] = 'cache.backend.memcache';
```

Comment:
Why would we not want to continue using APC for these bins?

Reply:
APC is enabled by default for these caches, but they are set to never expire so even when running a cr they are never purged which causes issues when you need to rebuild the service container - for example if you installed a views plugin and did a cache rebuild after deploying to 'refresh' a view using the new plugin, you'd often get a WSOD or other errors because there will be a null value due to the cache_discovery table having stale data. This snippet example was also intended to demonstrate the needed config to install Drupal with alternative cache bins by default and since this would require rebuilding the service container as modules / plugins were installed, this is not possible with the bootstrap / config / discovery using APCu because the stale cache objects often result in install errors.

Reply:
I need to revisit this, but is the cache lifetime related to the bin/cache entry, rather than the backend -- would these entries still be marked as `CACHE_PERMANENT` with memcache?

Reply:
in core the cache lifetime is related to the backend and then various bins use those backend / `cache_permanent` settings by default. in this cache the objects now using memcache as the backend would no longer use `CACHE_PERMANENT` and can be expired.

Reply:
+1 Awesome

Reply:
Also, very awesome that this includes the code change to cache Container on bootstrap.